compiler/src/iree/compiler/Codegen/LLVMGPU/TransformExtensions/LLVMGPUExtensions.cpp

```diff
@@ -815,5 +819,533 @@ DiagnosedSilenceableFailure transform_dialect::ReorderTransposeOp::applyToOne(
   return DiagnosedSilenceableFailure::success();
 }
 
+//===---------------------------------------------------------------------===//
+// EliminateGpuBarriersOp
+//===---------------------------------------------------------------------===//
+
+/// Returns `true` if the op is known not to have any side effects, but does not
+/// implement the MemoryEffectsOpInterface in the suitable way.
+static bool isKnownNoEffectsOpWithoutInterface(Operation *op) {
+  // memref::AssumeAlignment is conceptually pure, but marking it as such would
+  // make DCE immediately remove it.
+  return isa<memref::AssumeAlignmentOp>(op);
+}
+
+/// Returns `true` if the op is defines the parallel region that is subject to
+/// barrier synchronization.
+static bool isParallelRegionBoundary(Operation *op) {
+  if (op->hasAttr("__parallel_region_boundary_for_test")) return true;
+
+  // We consider functions inside executable variants that have the same symbol
+  // name as an export symbol.
+  auto func = dyn_cast<FunctionOpInterface>(op);
+  if (!func) return false;
+  auto parent = op->getParentOfType<ModuleOp>();
+  if (!parent) return false;
+  auto variant = parent->getParentOfType<HAL::ExecutableVariantOp>();
+  if (!variant) return false;
+  WalkResult result = variant.walk([&](HAL::ExecutableExportOp exportOp) {
+    if (exportOp.getSymNameAttr() == func.getNameAttr())
+      return WalkResult::interrupt();
+    return WalkResult::skip();
+  });
+  return result.wasInterrupted();
+}
+
+/// Returns `true` if the op behaves like a sequential loop, e.g., the control
+/// flow "wraps around" from the end of the body region back to its start.
+static bool isSequentialLoopLike(Operation *op) { return isa<scf::ForOp>(op); }
+
+/// Returns `true` if the regions of the op are guaranteed to be executed at
+/// most once. Thus, if an operation in one of the nested regions of `op` is
+/// executed than so are all the other operations in this region.
+static bool hasSingleExecutionBody(Operation *op) {
+  return isa<scf::IfOp, memref::AllocaScopeOp>(op);
+}
+
+/// Returns `true` if the operation is known to produce a pointer-like object
+/// distinct from any other object produced by a similar operation. For example,
+/// an allocation produces such an object.
+static bool producesDistinctBase(Operation *op) {
```

Comment:
Could this just be the Alloc effect or have you actually needed other cases in other contexts?

Reply:
There's a combination of potential false-positive and false-negative that makes me prefer a finer-grain approach here:
- false-positive: the conservative result of `getEffects` is to list all possible effects, including Alloc, but it's a just a way of saying "we don't really know anything about this op";
- false-negative: a `func.call` to `@malloc` also produces distinct base pointers, but function calls don't have an Alloc effect.